File: src/main/java/org/theosib/GraphicsEngine/Face.java
package org.theosib.GraphicsEngine;

import org.joml.Vector2f;
import org.joml.Vector2fc;
import org.joml.Vector3f;
import org.joml.Vector3fc;

import java.nio.FloatBuffer;

public interface Face {
    default public void setVertex(int index, float x, float y, float z) {
        setVertex(index, new Vector3f(x, y, z));
    }

    public void setVertex(int index, Vector3f point);

    public Vector3fc getVertex(int index);

    default public void setTexCoord(int index, float x, float y) {
        setTexCoord(index, new Vector2f(x, y));
    }

    public void setTexCoord(int index, Vector2f point);

    public Vector2fc getTexCoord(int index);

    public Vector3fc getNormal();

    public int numVertices();

    public int numTriangleVertices();

    default public int numVertexFloats() {
        return numTriangleVertices() * 3;
    }

    default public int numTextureFloats() {
        return numTriangleVertices() * 2;
    }

    default public int numNormalFloats() {
        return numVertexFloats();
    }

    default public void getVertexFloats(FloatBuffer outBuf, float offsetX, float offsetY, float offsetZ) {
        int loops = numTriangleVertices();
        for (int i = 0; i < loops; i++) {
            Vector3fc vertex = getVertex(i);
            outBuf.put(vertex.x() + offsetX);
            outBuf.put(vertex.y() + offsetY);
            outBuf.put(vertex.z() + offsetZ);
        }
    }

    default public void getTextcoordFloats(FloatBuffer outBuf) {
        int loops = numTriangleVertices();
        for (int i = 0; i < loops; i++) {
            Vector2fc tex = getTexCoord(i);
            outBuf.put(tex.x());
            outBuf.put(tex.y());
        }
    }

    default public void getNormalFloats(FloatBuffer outBuf) {
        int loops = numTriangleVertices();
        Vector3fc normal = getNormal();
        for (int i = 0; i < loops; i++) {
            outBuf.put(normal.x());
            outBuf.put(normal.y());
            outBuf.put(normal.z());
        }
    }

    static public class Triangle implements Face {
        private Vector3f[] vertices = new Vector3f[3];
        private Vector2f[] texcoords = new Vector2f[3];
        private Vector3f normal = null;

        @Override
        public void setVertex(int index, Vector3f point) {
            vertices[index] = point;
            normal = null;
        }

        @Override
        public Vector3fc getVertex(int index) {
            return vertices[index];
        }

        @Override
        public void setTexCoord(int index, Vector2f point) {
            texcoords[index] = point;
        }

        @Override
        public Vector2fc getTexCoord(int index) {
            return texcoords[index];
        }

        @Override
        public Vector3fc getNormal() {
            if (normal != null) return normal;

            Vector3f a = vertices[0];
            Vector3f b = vertices[1];
            Vector3f c = vertices[2];
            Vector3f cb = new Vector3f();
            Vector3f ab = new Vector3f();
            Vector3f x = new Vector3f();
            c.sub(b, cb);
            a.sub(b, ab);
            cb.cross(ab, x);
            normal = x.normalize();

            return normal;
        }

        @Override
        public int numVertices() {
            return 3;
        }

        @Override
        public int numTriangleVertices() {
            return 3;
        }
    }

    static public class Quad implements Face {
        private Vector3f[] vertices = new Vector3f[4];
        private Vector2f[] texcoords = new Vector2f[4];
        private Vector3f normal = null;
        private boolean solid = true;

        static private int[] repeat_indices = {
                0, 1, 2, 0, 2, 3
        };

        @Override
        public void setVertex(int index, Vector3f point) {
            vertices[index] = point;
            normal = null;
        }

        @Override
        public Vector3fc getVertex(int index) {
            return vertices[repeat_indices[index]];
        }

        @Override
        public void setTexCoord(int index, Vector2f point) {
            texcoords[index] = point;
        }

        @Override
        public Vector2fc getTexCoord(int index) {
            return texcoords[repeat_indices[index]];
        }

        @Override
        public Vector3fc getNormal() {
            if (normal != null) return normal;

            Vector3f a;
            Vector3f b;
            Vector3f c;

            if (vertices[0].equals(vertices[1]) || vertices[1].equals(vertices[2])) {
                a = vertices[0];
                b = vertices[2];
                c = vertices[3];
            } else {
                a = vertices[0];
                b = vertices[1];
                c = vertices[2];
            }

            Vector3f cb = new Vector3f();
            Vector3f ab = new Vector3f();
            Vector3f x = new Vector3f();
            c.sub(b, cb);
            a.sub(b, ab);
            cb.cross(ab, x);
            normal = x.normalize();

            return normal;
        }

        @Override
        public int numVertices() {
            return 4;
        }

        @Override
        public int numTriangleVertices() {
            return 6;
        }
    }

    static public class NoFace implements Face {
        public static final NoFace singleton = new NoFace();

        static final private Vector3f dummy3f = new Vector3f();
        static final private Vector2f dummy2f = new Vector2f();

        @Override
        public void setVertex(int index, Vector3f point) {
        }

        @Override
        public Vector3fc getVertex(int index) {
            return dummy3f;
        }

        @Override
        public void setTexCoord(int index, Vector2f point) {
        }

        @Override
        public Vector2fc getTexCoord(int index) {
            return dummy2f;
        }

        @Override
        public Vector3f getNormal() {
            return dummy3f;
        }

        @Override
        public int numVertices() {
            return 0;
        }

        @Override
        public int numTriangleVertices() {
            return 0;
        }
    }
}

File: junk/Face.scala
package org.theosib

import org.joml.{Vector2f, Vector2fc, Vector3f, Vector3fc}

import java.nio.FloatBuffer
import scala.collection.mutable.ListBuffer

class Face {
  val vertices = new ListBuffer[Vector3fc]
  val texcoords = new ListBuffer[Vector2fc]
  var normal: Vector3fc = null

  def numVertices(): Int = vertices.length

  def numTriangleVertices(): Int = {
    numVertices() match {
      case 0 => 0
      case 3 => 3
      case 4 => 6
      case _ => _
    }
  }

  def addVertex(x: Float, y: Float, z: Float): Face = {
    addVertex(new Vector3f(x, y, z))
  }

  def addVertex(vertex: Vector3fc): Face = {
    vertices.append(vertex)
    if (vertices.length >= 3) computeNormal()
    this
  }

  def addTexCoord(x: Float, y: Float): Face = {
    addTexCoord(new Vector2f(x, y))
  }

  def addTexCoord(texcoord: Vector2fc): Face = {
    texcoords.append(texcoord)
  }

  def computeNormal(): Unit = {
    var a: Vector3fc = null
    var b: Vector3fc = null
    var c: Vector3fc = null

    if (numVertices()==3) {
      a = vertices(0)
      b = vertices(1)
      c = vertices(2)
    } else {
      if (vertices(0).equals(vertices(1)) || vertices(0).equals(vertices(2))) {
        a = vertices(0)
        b = vertices(2)
        c = vertices(3)
      } else {
        a = vertices(0)
        b = vertices(1)
        c = vertices(2)
      }
    }

    val cb = new Vector3f()
    val ab = new Vector3f()
    c.sub(b, cb)
    a.sub(b, ab)
    val x = new Vector3f()
    cb.cross(ab, x)
    normal = x.normalize()
  }

  final private val face_indices = Array(0, 1, 2, 0, 2, 3)

  def getTriangleVertices(outBuf: FloatBuffer, start: Int, offset: Vector3fc): Int = {
    var index = start
    for (i <- 0 until numTriangleVertices()) {
      val v = face_indices(i)
      outBuf.put(index, vertices(v).x() + offset.x()); index += 1;
      outBuf.put(index, vertices(v).y() + offset.y()); index += 1;
      outBuf.put(index, vertices(v).z() + offset.z()); index += 1;
    }
    index
  }

  def getTriangleTextcoords(outBuf: FloatBuffer, start: Int): Int = {
    var index = start
    for (i <- 0 until numTriangleVertices()) {
      val v = face_indices(i)
      outBuf.put(index, texcoords(v).x()); index += 1;
      outBuf.put(index, texcoords(v).y()); index += 1;
    }
    index
  }

  def getTriangleNormals(outBuf: FloatBuffer, start: Int): Int = {
    var index = start;
    for (i <- 0 until numTriangleVertices()) {
      outBuf.put(index, normal.x()); index += 1;
      outBuf.put(index, normal.y()); index += 1;
      outBuf.put(index, normal.z()); index += 1;
    }
    index
  }

/*


   */
}
